java-design-patterns-master/event-asynchronous/src/main/java/com/iluwatar/event/asynchronous/EventManager.java

```diff
@@ -0,0 +1,221 @@
+   * @return eventId
+   * @throws MaxNumOfEventsAllowedException When too many events are running at a time.
+   * @throws InvalidOperationException      No new synchronous events can be created when one is
+   *                                        already running.
+   * @throws LongRunningEventException      Long running events are not allowed in the app.
+   */
+  public int create(int eventTime)
+      throws MaxNumOfEventsAllowedException, InvalidOperationException, LongRunningEventException {
+    if (currentlyRunningSyncEvent != -1) {
+      throw new InvalidOperationException("Event [" + currentlyRunningSyncEvent + "] is still"
+          + " running. Please wait until it finishes and try again.");
+    }
+
+    var eventId = createEvent(eventTime, true);
+    currentlyRunningSyncEvent = eventId;
+
+    return eventId;
+  }
+
+  /**
+   * Create an Asynchronous event.
+   *
+   * @param eventTime Time an event should run for.
+   * @return eventId
+   * @throws MaxNumOfEventsAllowedException When too many events are running at a time.
+   * @throws LongRunningEventException      Long running events are not allowed in the app.
+   */
+  public int createAsync(int eventTime) throws MaxNumOfEventsAllowedException,
+      LongRunningEventException {
+    return createEvent(eventTime, false);
+  }
+
+  private int createEvent(int eventTime, boolean isSynchronous)
+      throws MaxNumOfEventsAllowedException, LongRunningEventException {
+    if (eventPool.size() == MAX_RUNNING_EVENTS) {
+      throw new MaxNumOfEventsAllowedException("Too many events are running at the moment."
+          + " Please try again later.");
+    }
+
+    if (eventTime >= MAX_EVENT_TIME) {
+      throw new LongRunningEventException(
+          "Maximum event time allowed is " + MAX_EVENT_TIME + " seconds. Please try again.");
+    }
+
+    var newEventId = generateId();
+
+    var newEvent = new Event(newEventId, eventTime, isSynchronous);
+    newEvent.addListener(this);
+    eventPool.put(newEventId, newEvent);
+
+    return newEventId;
+  }
+
+  /**
+   * Starts event.
+   *
+   * @param eventId The event that needs to be started.
+   * @throws EventDoesNotExistException If event does not exist in our eventPool.
+   */
+  public void start(int eventId) throws EventDoesNotExistException {
+    if (!eventPool.containsKey(eventId)) {
+      throw new EventDoesNotExistException(eventId + DOES_NOT_EXIST);
+    }
+
+    eventPool.get(eventId).start();
+  }
+
+  /**
+   * Stops event.
+   *
+   * @param eventId The event that needs to be stopped.
+   * @throws EventDoesNotExistException If event does not exist in our eventPool.
+   */
+  public void cancel(int eventId) throws EventDoesNotExistException {
+    if (!eventPool.containsKey(eventId)) {
+      throw new EventDoesNotExistException(eventId + DOES_NOT_EXIST);
+    }
+
+    if (eventId == currentlyRunningSyncEvent) {
+      currentlyRunningSyncEvent = -1;
+    }
+
+    eventPool.get(eventId).stop();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **138 and 146**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **147** and **205**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).